### PR TITLE
twingate: init at 1.0.60

### PR DESCRIPTION
**Pull request by antonshkurenko**
###### Motivation for this change

Add a twingate package

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/networking/twingate/default.nix
Comment (SuperSandro2000) [inline pkgs/applications/networking/twingate/default.nix:18]:
```suggestion
    url = "https://binaries.twingate.com/client/linux/DEB/${version}/twingate-amd64.deb";
```

File: pkgs/applications/networking/twingate/default.nix
Comment (SuperSandro2000) [inline pkgs/applications/networking/twingate/default.nix:27]:
```suggestion
  postPatch = ''
```

File: pkgs/applications/networking/twingate/default.nix
Comment (SuperSandro2000) [inline pkgs/applications/networking/twingate/default.nix:51]: Please put this into a separate commit with the message ``maintainers: add NAME`` and sort it alphabetically.

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

File: nixos/modules/services/networking/twingate.nix
Comment (aanderse) [inline nixos/modules/services/networking/twingate.nix:12]:
```suggestion
    enable = mkEnableOption "Twingate Client daemon";

    settings = mkOption {
      type = (pkgs.formats.json {}).type;
      default = {};
      description = ''
        <insert description to this configuration file here - i couldn't find any references online>
      '';
    };

```

File: nixos/modules/services/networking/twingate.nix
Comment (aanderse) [inline nixos/modules/services/networking/twingate.nix:18]: When would this option ever be used? You should probably drop it.

File: nixos/modules/services/networking/twingate.nix
Comment (aanderse) [inline nixos/modules/services/networking/twingate.nix:81]: Drop this because it is already included when you wrote `systemd.packages = [ cfg.package ];`.

```suggestion
```

File: nixos/modules/services/networking/twingate.nix
Comment (aanderse) [inline nixos/modules/services/networking/twingate.nix:73]: If you drop the `package` option you don't need this (except for the `WantedBy` bit).
```suggestion
    systemd.services.twingate.wantedBy = [ "multi-user.target" ];
```

Why? https://github.com/NixOS/nixpkgs/issues/81138

FYI, so you know, if you want to override the `ExecStart` you do it like this:
```
    systemd.services.twingate.serviceConfig = [ "" "${cfg.package}/bin/twingated /etc/twingate/config.json" ];
```

File: nixos/modules/services/networking/twingate.nix
Comment (aanderse) [inline nixos/modules/services/networking/twingate.nix:32]: You don't need most of the files included in `${cfg.package}/etc/twingate`. And even if you did, you wouldn't want to copy over them every time.

```suggestion
    environment.etc."twingate/config.json".source = (pkgs.formats.json {}).generate "config.json" cfg.settings;
```

**Review by aanderse — CHANGES_REQUESTED** [review]
I have left some suggestions I hope you find useful. Please do not hesitate to ask questions or let me know if you need a hand. Always happy to help.

**Comment by tonyshkurenko** [comment]
Thanks for your comments!

File: nixos/modules/services/networking/twingate.nix
Comment (tonyshkurenko) [reply nixos/modules/services/networking/twingate.nix:32]: 1) Regarding copying every time: it's done on purpose
2) Regarding config.json: it isn't basically a config, we'd like to fill through nix files. Maybe in future, but for now it's more a static data, which shouldn't be changed a lot

**Comment by tonyshkurenko** [comment]
Please, re-review

File: nixos/modules/services/networking/twingate.nix
Comment (aanderse) [reply nixos/modules/services/networking/twingate.nix]: Are the files under `/etc/twingate` meant to be mutable or immutable? Is the directory `/etc/twingate` meant to be mutable or immutable? I'm having a bit of a hard time understanding the intended usage of this directory for your module. The `twingate` software merely reads from these files and doesn't write to them, correct?

Comment (tonyshkurenko) [reply nixos/modules/services/networking/twingate.nix]: @aanderse `/etc/twingate` should be mutable. We save user preferences there

Comment (aanderse) [reply nixos/modules/services/networking/twingate.nix]: @tonyshkurenko sorry this PR fell of my radar. So `twingate` manages its own configuration allowing you to modify it through its UI or something like that? I understand now... Thanks.

File: nixos/modules/services/networking/twingate.nix
Comment (aanderse) [inline nixos/modules/services/networking/twingate.nix:22]: Not required. `twingate.service` takes care of creating this directory via this line: `ConfigurationDirectory=twingate`.

File: nixos/modules/services/networking/twingate.nix
Comment (aanderse) [inline nixos/modules/services/networking/twingate.nix:24]: Please move this line into `systemd.services.twingate.preStart` instead.

File: maintainers/maintainer-list.nix
Comment (aanderse) [inline maintainers/maintainer-list.nix:11815]: Please keep indentation consistent at 2 spaces per level.

**Review by aanderse — CHANGES_REQUESTED** [review]
Could you also please rewrite your git history into 3 commits:
1. adding yourself to the maintainers list
2. adding the package
3. adding the nixos module

Thank you!

File: nixos/doc/manual/release-notes/rl-2111.section.md
Comment (SuperSandro2000) [inline nixos/doc/manual/release-notes/rl-2111.section.md:167]: This should contain where to find it.

File: nixos/modules/services/networking/twingate.nix
Comment (SuperSandro2000) [inline nixos/modules/services/networking/twingate.nix:29]: Missing final new line.

File: pkgs/applications/networking/twingate/default.nix
Comment (SuperSandro2000) [inline pkgs/applications/networking/twingate/default.nix:11]:
```suggestion
```

File: pkgs/applications/networking/twingate/default.nix
Comment (SuperSandro2000) [inline pkgs/applications/networking/twingate/default.nix:24]:
```suggestion
  nativeBuildInputs = [ dbus dpkg autoPatchelfHook ];

  buildInputs = [ curl libnl udev cryptsetup ];
```
Why do we need dbus in nativeBuildInputs?

File: pkgs/applications/networking/twingate/default.nix
Comment (SuperSandro2000) [inline pkgs/applications/networking/twingate/default.nix:33]: Are there many files that need to be patched? Usually patching should traget closely what needs patching and not be to broad.

**Comment by SuperSandro2000** [comment]
Please squash the commits.

File: pkgs/applications/networking/twingate/default.nix
Comment (alexprivalov) [reply pkgs/applications/networking/twingate/default.nix:32]: yes, that's required. most os configs contain pathes that should be changed

File: nixos/modules/services/networking/twingate.nix
Comment (alexprivalov) [reply nixos/modules/services/networking/twingate.nix:24]: Such change will break app logic

Comment (aanderse) [reply nixos/modules/services/networking/twingate.nix:24]: Can you please explain?

**Comment by antonshkurenko** [comment]
@SuperSandro2000 and @aanderse, please do re-review

File: nixos/modules/services/networking/twingate.nix
Comment (alexprivalov) [reply nixos/modules/services/networking/twingate.nix:24]: @aanderse , It seems that I mixed up something earlier. I've checked one more time and confirm that your suggestion is ok, applied.

**Comment by alexprivalov** [comment]
@aanderse what does this error stand?
image

**Comment by aanderse** [comment]
I have noticed documentation changing in some PRs lately but since I haven't been nearly as active in the last ~5 months I missed some of the details of this. Unfortunately I'm not sure the answer to your question, but I'm very curious to know what it is.

Anyone else on this PR: can you provide an explanation and maybe a link to some documentation of you're aware?

**Comment by sagie-shamay** [comment]
@SuperSandro2000 Can you please help us understand the problem in the NixOS manual build?

**Review by kalbasit — APPROVED** [review]

File: nixos/modules/services/networking/twingate.nix
Comment (alexprivalov) [reply nixos/modules/services/networking/twingate.nix]: deleted

File: nixos/modules/services/networking/twingate.nix
Comment (SuperSandro2000) [inline nixos/modules/services/networking/twingate.nix:11]:
```suggestion
    enable = mkEnableOption (lib.mdDoc "Twingate Client daemon");
```

Comment (alexprivalov) [reply nixos/modules/services/networking/twingate.nix]: @SuperSandro2000, thanks for suggestion!

**Comment by brodul** [comment]
My organization also uses `twingate` and I tried the PR.

I got an error 
image

Then I run `strace`

```
[...]
write(1, "\33[94mEnter the name of your Twin"..., 87Enter the name of your Twingate network (eg. "acme" for "acme.twingate.com"): ) = 87                                                    
read(0, myorg                                                                                                                                                                                
"myorg\n", 8192)                 = 5                                                                                                                                                         
write(1, "checking...\n", 12checking...                                                                                                                                                     
)           = 12                                                                                                                                                                            
brk(0x557c543d2000)                     = 0x557c543d2000                                                                                                                                    
brk(0x557c543be000)                     = 0x557c543be000                                                                                                                                    
openat(AT_FDCWD, "/etc/twingate/controller_url.conf", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)                                                                           
write(6, "<12>Oct 17 09:34:44 twingate-cli"..., 101) = 101                                                                                                                                  
mmap(NULL, 2101248, PROT_NONE, MAP_PRIVATE|MAP_ANONYMOUS|MAP_STACK, -1, 0) = 0x7feaeb598000                                                                                                 
mprotect(0x7feaeb599000, 2097152, PROT_READ|PROT_WRITE) = 0                                                                                                                                 
rt_sigprocmask(SIG_BLOCK, ~[], [], 8)   = 0                                                                                                                                                 
clone3({flags=CLONE_VM|CLONE_FS|CLONE_FILES|CLONE_SIGHAND|CLONE_THREAD|CLONE_SYSVSEM|CLONE_SETTLS|CLONE_PARENT_SETTID|CLONE_CHILD_CLEARTID, child_tid=0x7feaeb798910, parent_tid=0x7feaeb798
910, exit_signal=0, stack=0x7feaeb598000, stack_size=0x1ffd80, tls=0x7feaeb798640} => {parent_tid=[81593]}, 88) = 81593                                                                     
rt_sigprocmask(SIG_SETMASK, [], NULL, 8) = 0                                                                                                                                                
futex(0x557c54380330, FUTEX_WAIT_PRIVATE, 2, NULL) = 0                                                                                                                                      
write(6, "<15>Oct 17 09:34:44 twingate-cli"..., 75) = 75                                                                                                                                    
futex(0x557c54380330, FUTEX_WAKE_PRIVATE, 1) = 0                                                                                                                                            
write(8, "\1\0\0\0\0\0\0\0", 8)         = 8                                                                                                                                                 
write(6, "<15>Oct 17 09:34:44 twingate-cli"..., 57) = 57                                                                                                                                    
futex(0x557c54380330, FUTEX_WAKE_PRIVATE, 1) = 1                                                                                                                                            
write(6, "<15>Oct 17 09:34:44 twingate-cli"..., 81) = 81                                                                                                                                    
futex(0x557c54361b58, FUTEX_WAIT_PRIVATE, 4294967295, {tv_sec=29, tv_nsec=999996224}) = 0                                                                                                   
write(8, "\1\0\0\0\0\0\0\0", 8)         = 8                                                                                                                                                 
futex(0x7fead4000d30, FUTEX_WAKE_PRIVATE, 1) = 1                                                                                                                                            
write(6, "<15>Oct 17 09:34:44 twingate-cli"..., 77) = 77                                                                                                                                    
write(8, "\1\0\0\0\0\0\0\0", 8)         = 8                                                                                                                                                 
futex(0x557c54380330, FUTEX_WAIT_PRIVATE, 2, NULL) = -1 EAGAIN (Resource temporarily unavailable)                                                                                           
futex(0x557c54380330, FUTEX_WAIT_PRIVATE, 2, NULL) = -1 EAGAIN (Resource temporarily unavailable)                                                                                           
futex(0x557c54380330, FUTEX_WAIT_PRIVATE, 2, NULL) = -1 EAGAIN (Resource temporarily unavailable)                                                                                           
write(6, "<15>Oct 17 09:34:44 twingate-cli"..., 88) = 88                                                                                                                                    
futex(0x557c54380330, FUTEX_WAKE_PRIVATE, 1) = 1                                                                                                                                            
futex(0x7feaeb798910, FUTEX_WAIT_BITSET|FUTEX_CLOCK_REALTIME, 81593, NULL, FUTEX_BITSET_MATCH_ANY) = 0                                                                                      
write(6, "<15>Oct 17 09:34:44 twingate-cli"..., 76) = 76                                                                                                                                    
write(1, "\33[92mmyorg is a valid Twingate ne"..., 42myorg is a valid Twingate network                                                                                                        
) = 42                                                                                                                                                                                      
openat(AT_FDCWD, "/etc/twingate/network.conf", O_WRONLY|O_CREAT|O_TRUNC|O_CLOEXEC, 0666) = -1 ENOENT (No such file or directory)                                                            
write(2, "No such file or directory", 25No such file or directory) = 25                                                                                                                     
write(2, " (os error ", 11 (os error )             = 11                                                                                                                                     
write(2, "2", 12)                        = 1                                                                                                                                                
write(2, ")", 1))                        = 1                                                                                                                                                
write(2, "\n", 1                                                                                                                                                                            
)                       = 1                                                                                                                                                                 
write(1, "\33[91mExpected myorg but found ''."..., 55Expected myorg but found ''. Try again? [Y/n]: ) = 55
[...]
```

Figured out that the `twingate setup` fails if `/etc/twingate` folder is missing. Once added I got a connection.

**Comment by alexprivalov** [comment]
> My organization also uses `twingate` and I tried the PR.
> 
> I got an error image
> 
> ......
> 
> Figured out that the `twingate setup` fails if `/etc/twingate` folder is missing. Once added I got a connection.

@brodul Did you enable service(`services.twingate.enable = true;`) in /etc/nixos/configuration.nix?
Could you please share how did you setup twingate?

**Comment by alexprivalov** [comment]
@SuperSandro2000 and @aanderse, do you have any objections about PR?
Is it possible to merge PR?

**Review by kalbasit — APPROVED** [review]
I've been using Twingate for over a month now with no issues.

**Review by aanderse — APPROVED** [review]
Module looks fine :+1:

**Comment by kalbasit** [comment]
Please break the commit into 3 commits:

1. maintainers: add tonyshkurenko
2. twingate: init at 1.0.60
3. nixos/twingate: init module

I will merge once the commits are rebased.

File: nixos/modules/services/networking/twingate.nix
Comment (SuperSandro2000) [inline nixos/modules/services/networking/twingate.nix:17]: This needs a proper assertion. You can enable networkd and networkmanager but if done incorrect it can break your networking setup and I wouldn't expect that twingate turns on networkmanager.

Comment (kalbasit) [reply nixos/modules/services/networking/twingate.nix:17]: This PR has been running for some time now. I'm going to go ahead and merge it and let @antonshkurenko deal with this later.

**Pull request merged**
